invite: Add setup tips to user invite modal.

sayamsamal opened the pull request:
We include setup tips to the user invite modal for the following
cases:

(At the top of the invite modal)
- If the org description is missing.
- If the org profile picture is missing.
- If the custom profile fields have not been added.

We also use the new banner ui to display these tips. In doing so, we extract the banner component from `compose_banner.hbs` to `popover_banner.hbs`, removing any compose specific code from the banner component.

Fixes: #24262

Previous Work: #24273 

> [!NOTE]  
> - This PR includes the newer compose banner system as suggested in https://github.com/zulip/zulip/pull/24273#issuecomment-1716341552.
> - This PR includes setup tips pointing to default new user settings but does not actually check for any changes from the original default new user settings since we don't have a good way to fetch that information atm. Thus for the time being we merge the default new user setup tip with the custom profile fields setup tip.
See [CZO](https://chat.zulip.org/#narrow/stream/6-frontend/topic/Comparing.20User.20Default.20Settings.3F/near/1543272) and https://github.com/zulip/zulip/pull/25040
> - This PR doesn't include setup tips for when no new streams have been created, again due to the unavailability of a good way to check that. See [CZO](https://chat.zulip.org/#narrow/stream/101-design/topic/setup.20tips.20in.20user.20invite.20modal/near/1684370),


<!-- Describe your pull request here.-->

<!-- If the PR makes UI changes, always include one or more still screenshots to demonstrate your changes. If it seems helpful, add a screen capture of the new functionality as well.

Tooling tips: https://zulip.readthedocs.io/en/latest/tutorials/screenshot-and-gif-software.html
-->

**Screenshots and screen captures:**
![image](https://github.com/zulip/zulip/assets/82862779/d1a96a77-ae32-4528-88c4-2b7b65fabde0)


<details>
<summary>Self-review checklist</summary>

<!-- Prior to submitting a PR, follow our step-by-step guide to review your own code:
https://zulip.readthedocs.io/en/latest/contributing/code-reviewing.html#how-to-review-code -->

<!-- Once you create the PR, check off all the steps below that you have completed.
If any of these steps are not relevant or you have not completed, leave them unchecked.-->

- [x] [Self-reviewed](https://zulip.readthedocs.io/en/latest/contributing/code-reviewing.html#how-to-review-code) the changes for clarity and maintainability
      (variable names, code reuse, readability, etc.).

Communicate decisions, questions, and potential concerns.

- [x] Explains differences from previous plans (e.g., issue description).
- [x] Highlights technical choices and bugs encountered.
- [x] Calls out remaining decisions and concerns.
- [ ] Automated tests verify logic where appropriate.

Individual commits are ready for review (see [commit discipline](https://zulip.readthedocs.io/en/latest/contributing/commit-discipline.html)).

- [x] Each commit is a coherent idea.
- [x] Commit message(s) explain reasoning and motivation for changes.

Completed manual review and testing of the following:

- [x] Visual appearance of the changes.
- [x] Responsiveness and internationalization.
- [x] Strings and tooltips.
- [x] End-to-end functionality of buttons, interactions and flows.
- [x] Corner cases, error conditions, and easily imagined bugs.
</details>

alya commented:
Nice!

Just from the screenshots, the alignment of the `X` button looks too low to me in the two-line banners. (See compose banners for reference.)

alya commented:
When I make a brand new organization in dev, I don't see this message, and I should:

> If org description is missing: "You may want to [configure](https://github.com/zulip/zulip/issues/24262) your organization's login page prior to inviting users."

sayamsamal commented:
Pushed changes after discussion over at [CZO](https://chat.zulip.org/#narrow/stream/101-design/topic/setup.20tips.20in.20user.20invite.20modal/near/1683824).

I have updated the PR description to reflect the same and here is the updated screenshot for a new organisation.
![image](https://github.com/zulip/zulip/assets/82862779/dfd1bd1d-57c7-4f01-967c-31b53052eebf)

sayamsamal commented:
Changes in the updated code:
1. Fixed the org description missing banner not showing, this was due to my mistake of not converting `or` to `and` when reversing the logic from check if the description doesn't exist to check if the description exists.
2. Merged `default new user settings` and `custom profile fields` banner, since we don't have a good way to check for changes in `default new user settings` atm.
3. Removed the `no streams created` banner, due to reasons similar to the above.

alya commented:
Works for me! I think this is a fine checkpoint, and we can consider the rest of this issue as follow-ups.

@sahil839 perhaps you could review this one? It's a release goal.

sahil839 commented on web/src/invite.ts:
```diff
@@ -288,6 +290,18 @@ function set_streams_to_join_list_visibility(): void {
+            realm_description !== "" &&
+            !/^Organization imported from [A-Za-z]+[!.]$/.test(realm_description),
+        is_profile_pic: realm_icon_source !== "G",
+        is_custom_profile: custom_profile_fields.length > 0,
```
I think these variables can have better names like - `is_realm_description_set` or maybe `is_realm_description_configured`,  `is_realm_icon_user_uploaded` or `is_realm_profile_picture_user_uploaded`, `realm_has_custom_profile_fields`.

Basically, it would be better to use `custom_profile_fields` instead of `custom_profile` and `realm_icon` instead of `profile_pic` and be a little bit verbose to explain the use of variable.

sayamsamal replied:
I have renamed the variables to: `realm_has_description`, `realm_has_user_set_icon` and `realm_has_custom_profile_fields` to balance the variable length vs verbosity. Does this sound good to you?

sahil839 commented on web/src/invite.ts:
```diff
@@ -329,6 +343,15 @@ function open_invite_user_modal(e: JQuery.ClickEvent<Document, undefined>): void
+
+        $("#invite-user-modal").on("click", `.main-view-banner-close-button`, (event) => {
+            event.preventDefault();
+            $(event.target).parent().remove();
```
Not a major thing, but we usually use `e` instead of `event` in such cases.

sayamsamal replied:
Fixed, thanks!

sahil839 commented on web/src/invite.ts:
```diff
@@ -329,6 +343,15 @@ function open_invite_user_modal(e: JQuery.ClickEvent<Document, undefined>): void
+            dialog_widget.close();
+        });
+
+        $("#invite-user-modal").on("click", `.main-view-banner-close-button`, (event) => {
```
Why `main-view-banner-close-button` is enclosed in backticks here and not inside quotes like we do elsewhere?

sayamsamal replied:
Fixed and my bad for overlooking this. Thanks!

sahil839 commented on web/templates/invite_user_modal.hbs:
```diff
@@ -1,4 +1,9 @@
+    <div class="setup-tips-container {{#unless (or is_admin is_owner)}}hide{{/unless}}">
+        <div class="setup-tips-org-desc"></div>
+        <div class="setup-tips-org-profile-pic"></div>
+        <div class="setup-tips-custom-profiles"></div>
```
If we want to show only one banner at a time, can we just have a single div with a class name like "organization-setup-tips`?

alya replied:
I think we want to show multiple banners, potentially (one banner for org description / profile pic + potentially a banner for custom profile fields).

sayamsamal replied:
As mentioned by @alya we need to conditionally show the banners over here and found using different `divs` to be the simplest method of conditionally rendering and then removing the banners.

sahil839 reviewed (commented):
Posted a few comments.

sayamsamal commented on web/templates/invite_user_modal.hbs:
```diff
@@ -1,4 +1,9 @@
 <form id="invite-user-form">
+    <div class="setup-tips-container {{#unless (or is_admin is_owner)}}hide{{/unless}}">
```
Removed the check for `is_owner`, thanks for pointing this out!

sayamsamal commented:
@sahil839 I have tried my best to incorporate your feedback. Could you please have another look at this? Thanks!

sahil839 commented:
Looks good to me. @alya I guess you have taken a look and we can mark this ready for review by @timabbott. I have updated the labels.

timabbott commented on web/src/invite.ts:
```diff
@@ -415,6 +438,66 @@ function open_invite_user_modal(e: JQuery.ClickEvent<Document, undefined>): void
+                        )}</a>`,
+                },
+            );
+            $("#invite-user-form .setup-tips-container .setup-tips-custom-profiles").html(
```
Did you consider using the approach we take in `web/templates/compose_banner/automatic_new_visibility_policy_banner.hbs`? That's our intended approach for HTML banners, because having Handlebars do all the rendering via the `partial-block` syntax protects us from potential XSS security bugs that one risks with passing around HTML using the `{{{` syntax.

I don't think this specific PR would actually be a security bug, but we try hard as a project to avoid adding constructions that require paying attention to avoid XSS bugs.

sayamsamal replied:
I had tried that approach but couldn't get it to work due to some issues with not being able to refer to the `compose_banner` partial present in the `templates/compose_banner` from the parent directory (`templates`). I'll try again to see if I can get it working or start a discussion at CZO for the same. Thanks!

timabbott replied:
OK, yeah let's discuss in #frontend.

timabbott commented:
Posted one comment, otherwise lgtm!

timabbott commented:
Dropping "integration review" label just to clean up my tracker; please re-add it when you've addressed the above.

sayamsamal commented:
@timabbott In order to address https://github.com/zulip/zulip/pull/27464#discussion_r1401040547, I have extracted the relevant logic of `web/templates/compose_banner/compose_banner.hbs` to a new `web/templates/popover_banner/popover_banner.hbs`, stripping away the compose related logic included in that file. The idea behind that is, we can use this new partial for more generic banners. Now, the working of the `web/templates/popover_banner/invite_tips_banner.hbs` is very similar to that of `web/templates/compose_banner/automatic_new_visibility_policy_banner.hbs`.

In a future PR, it would make sense to extract the banner logic from the compose banner to a more common place, so that we can avoid having to duplicate the same template again and again.

sayamsamal commented:
@zulipbot add "integration review"

timabbott commented:
I renamed your new template to `modal_banner`, since popovers are our menus -- these are modals.

timabbott commented:
See also https://chat.zulip.org/#narrow/stream/101-design/topic/setup.20tips.20in.20user.20invite.20modal/near/1685627.
